tiledb/api/src/query/conditions.rs

```diff
@@ -0,0 +1,754 @@
+            op: CmpOp::NotEqual,
+            value: value.into(),
+        }))
+    }
+
+    pub fn ge<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::GreaterEqual,
+            value: value.into(),
+        }))
+    }
+
+    pub fn gt<V: Into<CmpValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::Cmp(CmpCondition {
+            field: self.field,
+            op: CmpOp::Greater,
+            value: value.into(),
+        }))
+    }
+
+    pub fn is_in<V: Into<SetMemberValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::SetMember(SetMemberCondition {
+            field: self.field,
+            op: SetMemberOp::In,
+            value: value.into(),
+        }))
+    }
+
+    pub fn not_in<V: Into<SetMemberValue>>(self, value: V) -> QC {
+        QC::Cond(Condition::SetMember(SetMemberCondition {
+            field: self.field,
+            op: SetMemberOp::NotIn,
+            value: value.into(),
+        }))
+    }
+
+    pub fn is_null(self) -> QC {
+        QC::Cond(Condition::Null(NullnessCondition {
+            field: self.field,
+            op: NullnessOp::IsNull,
+        }))
+    }
+
+    pub fn not_null(self) -> QC {
+        QC::Cond(Condition::Null(NullnessCondition {
+            field: self.field,
+            op: NullnessOp::NotNull,
+        }))
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum QC {
+    Cond(Condition),
+    Comb {
+        lhs: Box<QC>,
+        rhs: Box<QC>,
+        op: CombineOp,
+    },
+    Negate(Box<QC>),
+}
+
+impl QC {
+    pub fn field<F: AsRef<str>>(field: F) -> Field {
+        Field {
+            field: field.as_ref().to_owned(),
+        }
+    }
+
+    pub fn build<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        match self {
+            Self::Cond(cond) => cond.build(ctx),
+            Self::Comb { lhs, rhs, op } => {
+                let lhs = lhs.build(ctx)?;
+                let rhs = rhs.build(ctx)?;
+
+                let mut c_cond: *mut ffi::tiledb_query_condition_t = out_ptr!();
+                ctx.capi_return(unsafe {
+                    ffi::tiledb_query_condition_combine(
+                        ctx.capi(),
+                        lhs.capi(),
+                        rhs.capi(),
+                        op.capi_enum(),
+                        &mut c_cond,
+                    )
+                })?;
+
+                Ok(QueryCondition {
+                    context: ctx,
+                    raw: RawQueryCondition::Owned(c_cond),
+                })
+            }
+            Self::Negate(expr) => {
+                let mut c_cond: *mut ffi::tiledb_query_condition_t = out_ptr!();
+                ctx.capi_return(unsafe {
+                    ffi::tiledb_query_condition_negate(
+                        ctx.capi(),
+                        expr.build(ctx)?.capi(),
```

Comment:
`expr.build(ctx)` is not unsafe, I think you should pull this out of the `unsafe` block.
It also will help future readers who ask the question "wait, don't we have to build the `expr` first? Where is that? Oh wait there it is"

Reply:
As a generalization of this, I think most of our code kind of trains readers to ignore most of the FFI arguments because they are generally just simple `let` bindings.  I don't think we should deviate that

Reply:
That's fair. Will update.

Reply:
So, that point actually made me come around to pre-declaring all of the FFI arguments outside of unsafe (other than calling `as_ptr()` on anything related to `cstring!(foo)`). So I'll definitely keep doing that going forward.

Also, as I went to move that line, I suddenly realized I think that was actually a super subtle bug that we might want to think about checking things in CI with ASAN. I'll check if that's an easy thing to turn on. But basically, I'm pretty sure that was a "returning pointer to temporary that doesn't live long enough" but Rust can't tell. The `build(ctx)?` return value is a temporary there, and then immediately calling `capi()` on it means that it the actual underlying C++ query condition would get deleted before the pointer is passed to FFI. I'm not sure how drop works in unsafe blocks so I'm not sure if we were being saved by something specific there, or perhaps we just got lucky that C++ could read the bit of memory due to shenanigans.

Either way, the result of `build` is now a temporary let assignment. I just wonder if there's something we could do to tell the compiler that the results of capi calls has to live shorter than the object. Do you think pointers can have lifetimes? Looks like the closest we can get would be to return our `RawThing` structs and then deref them since it seems Rust won't do lifetimes on pointers: https://stackoverflow.com/a/27940991